crates/apollo-compiler/src/validation/operation.rs

```diff
@@ -1,51 +1,117 @@
+use crate::collections::HashSet;
 use crate::executable;
+use crate::validation::diagnostics::DiagnosticData;
 use crate::validation::DiagnosticList;
 use crate::validation::ExecutableValidationContext;
+use crate::validation::RecursionLimitError;
 use crate::ExecutableDocument;
+use crate::Name;
 use crate::Node;
 
+/// Iterate all selections in the selection set.
+///
+/// This includes fields, fragment spreads, and inline fragments. For fragments, both the spread
+/// and the fragment's nested selections are reported.
+///
+/// Does not recurse into nested fields.
+fn walk_selections<'doc>(
+    document: &'doc ExecutableDocument,
+    selections: &'doc executable::SelectionSet,
+    mut f: impl FnMut(&'doc executable::Selection),
+) -> Result<(), RecursionLimitError> {
+    fn walk_selections_inner<'doc>(
+        document: &'doc ExecutableDocument,
+        selection_set: &'doc executable::SelectionSet,
+        seen: &mut HashSet<&'doc Name>,
+        f: &mut dyn FnMut(&'doc executable::Selection),
+    ) -> Result<(), RecursionLimitError> {
```

Comment:
The implementation here doesn't actually raise a RecursionLimitError right now, but that will be part of a follow-up PR